src/roadmap.md

```diff
@@ -1,198 +1,192 @@
+description. As a result, services will be described uniformly, thus, making
+clear which transactions and arguments they process.
+
+Uniform service definition will also ease the development of light clients in
+any language supported by Protobuf. In the first phase of the implementation,
+service interfaces will support a single method type – transactions. Other
+method types will be added in one of the following versions.
+
+### Service Identification
+
+We plan on implementing a more logical mechanism of service identification in
+Exonum. Such a solution will ease the communication between services and
+external clients. This mechanism will also lay the foundation for the
+implementation of dynamic services.
+
+Within the scope of implementing the service identification mechanism, we also
+plan to provide a uniform framework for installing external services in
+Exonum blockchains.
+
+### Secure Workflow for Node Administration
+
+The introduction of a workflow for administrative operations, such as the
+deployment of services, is aimed at improving the security of the system.
+The exact details of the workflow are being currently discussed. One of the
+promising ideas is devising a service through which all the administrative
+processes will be conducted.
+
+### Secure Storage for Private Keys of the Node
+
+A separate and secured storage for private keys of the nodes will enhance the
+safety of the
+system. Even in the case of unauthorized users gaining access to a node,
+they will not have access to its private key and will not be able to
+intervene into the consensus process.
+
+### Java Binding: Protobuf Support and Separation of Transactions and Messages
+
+As we develop Java Binding alongside with Exonum Core, to equalize their
+functionality it is required to implement Protobuf support and separate the
+transactions and messages for Java Binding too. This step will promote all the
+corresponding [benefits](#Protobuf-as-the-Serialization-Format-in-Exonum)
+discussed above.
+
+## First Quarter of 2019
+
+### Service Interface Standard: Read Requests
+
+To increase the performance of the system, we will add read requests to the
+list of methods supported by interfaces of services. Read requests do not modify
+the blockchain state and can be handled locally on the full node which
+receives the request from a client.
+
+### Pre-check of Transactions in the Memory Pool
+
+Checking the validity of transactions which are included into the memory pool
+is one of the methods for preventing DoS attacks. This check will also ensure
+that transactions do not lead to any errors when processed.
+
+## Second Quarter of 2019
+
+### Dynamic Java Services
+
+The introduction of dynamic Java services will enable adding Java services to
+a running blockchain without the need to restart nodes. New services, in the
+form of Java modules, will be included into the network on the go. Support for
```

Comment:
I am not sure "in the form of Java modules" belongs here because Java modules might mean specific things (i.e., JPMS modules). The actual form they will be defined and packaged in is TBD.